src/automaton.cpp

```diff
@@ -0,0 +1,484 @@
+	X(Semicolon, SEMICOLON, ";")                                               \
+	X(Comma, COMMA, ",")                                                       \
+	X(Dot, DOT, ".")                                                           \
+	X(At, AT, "@")                                                             \
+                                                                               \
+	X(LParen, PAREN_OPEN, "(")                                                 \
+	X(RParen, PAREN_CLOSE, ")")                                                \
+	X(LBrace, BRACE_OPEN, "{")                                                 \
+	X(RBrace, BRACE_CLOSE, "}")                                                \
+	X(LBracket, BRACKET_OPEN, "[")                                             \
+	X(RBracket, BRACKET_CLOSE, "]")                                            \
+	X(LPoly, POLY_OPEN, "<:")                                                  \
+	X(RPoly, POLY_CLOSE, ":>")                                                 \
+                                                                               \
+	X(Identifier, END, {})                                                     \
+	X(String, STRING, {})                                                      \
+	X(Comment, END, {})                                                        \
+	X(Integer, INTEGER, {})                                                    \
+	X(Number, NUMBER, {})
+
+#define X(name, token_tag, string) name,
+enum Values { Error, END_STATES Count };
+#undef X
+#define X(name, token_tag, string) #name,
+constexpr char const* end_states[] = { "Error", END_STATES };
+#undef X
+constexpr int last_fixed_string = RPoly;
+}
+#define X(name, token_tag, string) string,
+static InternedString fixed_strings[] = { END_STATES };
+#undef X
+#define X(name, token_tag, string) TokenTag::token_tag,
+constexpr TokenTag token_tags[] = { END_STATES };
+#undef X
+#undef END_STATES
+
+constexpr void init_transitions(Automaton& result) {
+	// TODO: CLRF, more error handling
+
+	using namespace EndStates;
+	int normal_states = state_count;
+
+#define new_state() (--normal_states)
+
+	for (int i = state_count; i--;)
+		add_default_transition(result, i, Error);
+
+	State start = new_state(); // state_count - 1
+
+	// Identifiers
+	State saw_id_char = new_state();
+	add_transition(result, start, '_', saw_id_char);
+	for (char c = 'a'; c <= 'z'; ++c) add_transition(result, start, c, saw_id_char);
+	for (char c = 'A'; c <= 'Z'; ++c) add_transition(result, start, c, saw_id_char);
+	add_default_transition(result, saw_id_char, Identifier);
+	add_transition(result, saw_id_char, '_', saw_id_char);
+	for (char c = 'a'; c <= 'z'; ++c) add_transition(result, saw_id_char, c, saw_id_char);
+	for (char c = 'A'; c <= 'Z'; ++c) add_transition(result, saw_id_char, c, saw_id_char);
+	for (char c = '0'; c <= '9'; ++c) add_transition(result, saw_id_char, c, saw_id_char);
```

Comment:
`RANGE('a', 'z')` macro is great for this. `add_self_transition(state, char)` is very helpful for readability.

Reply:
I'm not sure about RANGE macro. I like the helper, I'm going with `add_loop`, though.